### PR TITLE
Mesh volume conservation

### PR DESCRIPTION
**Pull request by SchoeniPhlippsn**
## Description

This adds a global volume constraint potential which can be applied to the mesh data structure.

## Motivation and context

This volume constraint allows applying a global target volume per mesh type to closed membranes represented by a particle mesh.

## How has this been tested?

I added pytests and used this potential in my current studies.

## Change log

```
Addition of a volume constraint mesh potential
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: sphinx-doc/module-md-mesh-conservation.rst [hunk 1]
```diff
@@ -0,0 +1,22 @@
+.. Copyright (c) 2009-2022 The Regents of the University of Michigan.
+.. Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+md.mesh.conservation
+------------
```
Comment (tommy-waltmann) [inline sphinx-doc/module-md-mesh-conservation.rst]:
```suggestion
md.mesh.conservation
--------------------
```
I think this is why the documentation is failing to build properly.

**Comment by SchoeniPhlippsn** [comment]
pre-commit.ci autofix

**Comment by SchoeniPhlippsn** [comment]
pre-commit.ci autofix

**Comment by SchoeniPhlippsn** [comment]
@joaander @Charlottez112 
Ready for review!

File: hoomd/md/VolumeConservationMeshForceCompute.cc [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+
+    unsigned int n_types = m_mesh_data->getMeshTriangleData()->getNTypes();
+
+    if (m_ignore_type)
+        n_types = 1;
+
+    GPUArray<Scalar2> params(n_types, m_exec_conf);
+    m_params.swap(params);
+
+    m_volume = new Scalar[n_types];
```
Comment (joaander) [inline hoomd/md/VolumeConservationMeshForceCompute.cc]: `m_volume` is duplicated as `m_volume_GPU` in the subclass. Use `GPUArray<Scalar> m_volume` in the base class and then there is no need to duplicate code.

File: hoomd/md/mesh/conservation.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+    :py:class:`Volume` specifies a volume constraint on the whole mesh
+    surface:
+
+    .. math::
+
+        U(r) = k \frac{( V(r) - V_0 )^2}{2 \cdot V_0}
+
+    Args:
+        mesh (:py:mod:`hoomd.mesh.Mesh`): Mesh data structure constraint.
```
Comment (joaander) [inline hoomd/md/mesh/conservation.py:37]: Document `ignore_type`

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! I have a some suggestions for improvement. You will also need to merge `trunk-minor` with this branch.

**Comment by SchoeniPhlippsn** [comment]
pre-commit.ci autofix

**Comment by SchoeniPhlippsn** [comment]
pre-commit.ci autofix

**Review by joaander — APPROVED** [review]
Thanks!

**Pull request merged**